### ISSUE TITLE
System.Runtime.Caching is not supported on this platform

### ISSUE THREAD

**Issue by SidShetye** [Azure/azure-functions-host#4581]
We publish .NET standard libraries that use `MemoryCache` from `System.Runtime.Cache`. One of our customers wants to use our libraries with Azure Functions but this doesn't work. The error is 

`System.Private.CoreLib: ... <snip> ... System.Runtime.Caching: System.Runtime.Caching is not supported on this platform.`

Interestingly `MemoryCache` already works with .NET Core 2.1 ([info](https://github.com/dotnet/corefx/issues/14529#issuecomment-406668169)) but Azure Functions 2.0 (based on .NET Core 2.1) doesn't support this. This issue was also [previously reported](https://github.com/Azure/azure-functions-host/issues/3155) but the reporter closed it simply by changing user code, something not feasible for library publishers.

This is an urgent concern because as library publishers, we want to follow good software engineering and project management principles to support our .NET customers in a general way like .NET Standard or .NET Core versus highly specific combinations like .NET Core + Azure Functions etc that only leads to feature fragmentation and customer confusion.

In summary, please support `MemoryCache` from `System.Runtime.Cache` out of the box. In the meanwhile, is there any configuration or steps to manually add support for it in a given Azure Function App?

### UPDATE

TL;DR The fix in [comment](https://github.com/Azure/azure-functions-host/issues/4581#issuecomment-510630802) works. Just add this to the project `csproj` file till azure functions' tools handle this in the future.

```
  <Target Name="PostBuild" AfterTargets="PostBuildEvent">
    <Exec Command="copy $(OutDir)$(ProjectName).deps.json $(OutDir)bin\function.deps.json" />
  </Target>
  <Target Name="PostPublish" AfterTargets="AfterPublish">
    <Exec Command="copy $(PublishDir)$(ProjectName).deps.json $(PublishDir)bin\function.deps.json" />
  </Target>
```

#### Investigative information

Please provide the following:

- Timestamp: 6/17/2018
- Function App version (1.0 or 2.0): 2.0
- Function App name: doesn't matter
- Function name(s) (as appropriate): doesn't matter
- Invocation ID: doesn't matter
- Region: doesn't matter

**Comment by PureKrome** [comment]
@SidShetye Just thinking off the top of my head, but instead of using a `MemoryCache` is there any chance you could use `IMemoryCache` ([docs here](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.caching.memory.imemorycache?view=aspnetcore-2.1)) which are part of [`Microsoft.Extensions.Caching.Abstractions' nuget package](https://www.nuget.org/packages/Microsoft.Extensions.Caching.Abstractions/3.0.0-preview6.19304.6).

This way, you customers can choose which caching implementation to plug in.

**Comment by SidShetye** [comment]
@PureKrome Regardless of the details, it still means library publishers must rewrite to consider Azure Functions' limitations. It also adds configuration and documentation complexity instead of having things "just work". 

@mathewc or @fabiocav - what do you guys think? With support for `System.Runtime.Cache` already in .NET Core 2.1, can you support this in Azure Functions relatively easily? Also, just to add, our use for the cache isn't for anything ASP.NET related - just pure generic .NET Core business logic.

**Comment by SidShetye** [comment]
To add, even ASP.NET [documentation](https://docs.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-3.0) says `System.Runtime.Cache` should work, so having Azure Functions support it would keep customer experience consistent.

**Comment by PureKrome** [comment]
@SidShetye 

> Regardless of the details, it still means library publishers must rewrite to consider Azure Functions' limitations. It also adds configuration and documentation complexity instead of having things "just work".

Can you please explain what you mean by this? If I was going to be adding some internal caching into a nuget package (which also has nothing to do with ASP.NET stuff with respect to caching), I would still be using `IMemoryCache` and this is generic across _anything_ .NET Core related. So I wouldn't care if your host would be a console-app, web-app or functions-app ... it's agnostic to the app.

So this means there is no 'special Azure Function nuget package version'.

Also - why do you have to use `System.Runtime.Cache`? Again you're saying:

> Also, just to add, our use for the cache isn't for anything ASP.NET related - just pure generic .NET Core business logic.

which means `IMemoryCache` is a great .NET Core candidate for this problem? 

Are you really determined to stick with `System.Runtime.Cache`? Is there a reason for this?

**Comment by fabiocav** [comment]
Agree with what @PureKrome has stated above. Is what he proposed an adequate solution for you? If not, can you share more about details? Also, have you tried this on more recent builds?

This issue is actually something a tooling update will be addressing, as the runtime probing behavior is identical to what you'll experience in .NET Core.

For now, you can simply add the following to your project file and you should be able to use that dependency:

``` xml
  <Target Name="PostBuild" AfterTargets="PostBuildEvent">
    <Exec Command="copy $(OutDir)$(ProjectName).deps.json $(OutDir)bin\function.deps.json" />
  </Target>
  <Target Name="PostPublish" AfterTargets="AfterPublish">
    <Exec Command="copy $(PublishDir)$(ProjectName).deps.json $(PublishDir)bin\function.deps.json" />
  </Target>
```

**Comment by SidShetye** [comment]
@PureKrome Everything you're suggesting required change to library source code and updated to end customer application. That's a big effort for legacy/existing binaries. Even disregarding that and side-stepping any app-layer architecture dialog, this limitation simply violates the separation of concern. There is no reason the server host (azure functions) should purposely disallow a library (System.Runtime.Caching) otherwise allowed by the application runtime (.net core 2.0). At least I haven't seen a good reason.

> ... as the runtime probing behavior is identical to what you'll experience in .NET Core ...

@fabiocav This isn't a .NET Core issue since everything works just fine on .NET Core 2.0 - it's only when running this on Azure functions that the problem surfaces. This "not allowed" is really strange - why is this even not allowed in azure functions when the underlying runtime has no issues with the library? 

What other .NET Core libraries will be arbitrarily blacklisted from the azure functions hosting service?

**Comment by fabiocav** [comment]
Sorry, I don't think I was clear above. We don't block any .NET Core libraries (some APIs may be blocked in the sandbox when hosted, but that doesn't apply to this). The tooling issue I've mentioned is the Azure Functions SDK build/tools, not .NET Core tooling.

Have you tried adding the snippet I shared above to your project? You'll find that you'll be able to run without any modifications to your code.

**Comment by SidShetye** [comment]
@fabiocav Ok, this makes a lot more sense. Thanks for the clarity.

For closure with other engineers dealing with this issue, I quickly tried your patch to the `.csproj` in a new visual studio Azure Functions default template and can confirm that the patch **does work** and `System.Runtime.Caching` works just fine. So that's good news.

Before closing this item, is there an bug report on azure functions tools for this you can add here? Will automatically answer questions on which release train will get the fix.

**Comment by fabiocav** [comment]
@SidShetye you can use this issue to track the fix: https://github.com/Azure/azure-functions-vs-build-sdk/issues/271

Thank you for validating the workaround.

**Comment by PureKrome** [comment]
Nice work, awesome peeps!

**Comment by thecontrarycat** [comment]
@fabiocav I'm having a problem where the workaround only works locally.

That is to say,

- Running the project locally will fail unless the workaround is in place
- A published functions app *in Azure always fails*
- Downloading the files from that functions app from Azure and then running locally using the functions host *will work*

Is anyone else seeing this?

**Comment by thecontrarycat** [comment]
Oh, interesting.  It seems that if you deploy using *Run From Package*, the workaround *doesn't work*, but if you use *Web Deploy*, it *does*.

**Comment by kashimiz** [comment]
@fabiocav I have a case where the workaround isn't working locally. Please see issue #4795; repro solution is at [kashimiz/functions-memorycache-repro](https://github.com/kashimiz/functions-memorycache-repro).

**Comment by leniency** [comment]
> @fabiocav I'm having a problem where the workaround only works locally.
> 
> That is to say,
> 
> * Running the project locally will fail unless the workaround is in place
> * A published functions app _in Azure always fails_
> * Downloading the files from that functions app from Azure and then running locally using the functions host _will work_
> 
> Is anyone else seeing this?

I am. The work around was working fine last week. Then redeployed (no major changes) and keeps throwing errors about System.Drawing that worked before.

**Comment by leniency** [comment]
> @fabiocav I'm having a problem where the workaround only works locally.
> 
> That is to say,
> 
> * Running the project locally will fail unless the workaround is in place
> * A published functions app _in Azure always fails_
> * Downloading the files from that functions app from Azure and then running locally using the functions host _will work_
> 
> Is anyone else seeing this?

I got mine working again with one extra change. It looks like there was an additional /runtimes/ folder being set at the root level of the package, but should have also been in the bin folder.

I added an xcopy to move it on publish and it's working again.

`<Target Name="PostBuild" AfterTargets="PostBuildEvent">
    <Exec Command="copy $(OutDir)$(ProjectName).deps.json $(OutDir)bin\function.deps.json" />
  </Target>
  <Target Name="PostPublish" AfterTargets="AfterPublish">
    <Exec Command="copy $(PublishDir)$(ProjectName).deps.json $(PublishDir)bin\function.deps.json" />
    <Exec Command="xcopy $(PublishDir)runtimes $(PublishDir)bin\runtimes\ /e" />
  </Target>`
